src/enumerable.cr

```diff
@@ -609,6 +609,37 @@ module Enumerable(T)
     memo
   end
 
+  # Similar to `reduce`, but instead of raising when the input is empty,
+  # return `nil`
+  #
+  # ```
+  # ([] of Int32).reduce? { |acc, i| acc + i } # => nil
+  # ```
+  def reduce?
+    memo = uninitialized T
+    found = false
+
+    each do |elem|
+      memo = found ? (yield memo, elem) : elem
+      found = true
+    end
+
+    found ? memo : nil
+  end
+
+  # Similar to `reduce?`, however instead of returning `nil` when the input is empty,
+  # return the initial value of the accumulator.
```

Comment:
From the docs and tests for `reduce(memo)`, it appears that returning the `memo` is desired rather than `raise` https://github.com/crystal-lang/crystal/pull/4828/files#diff-6519b36b7135a1958e191f91f98354d8L560